kythe/docs/schema/indexing-protobuf.txt

```diff
@@ -31,12 +31,235 @@ implementation, as this is often difficult (or uninteresting) to read. This
+[source,javascript]
+-------------------------------------------------------------------------------
+const Color = {
+  RED: 0,
+};
+-------------------------------------------------------------------------------
+
+==== Changes to Generator
+
+To support cross-references betwen `color.js` and `Color.java` we need to update
+Generator to output the following mapping data for `Color`, `RED` and `GREEN`
+elements.
+
+[source,json]
+-------------------------------------------------------------------------------
+{
+    "type": "kythe0",
+    "meta": [{
+        "type": "anchor_anchor",
+        "source_begin": 13,
+        "source_end": 18,
+        "target_begin": 6,
+        "target_end": 11,
+        "edge": "/kythe/edge/imputes",
+        "source_vname": {
+            "corpus": "corpus",
+            "path": "path/to/Color.java"
+        }
+    }, {
+        "type": "anchor_anchor",
+        "source_begin": 22,
+        "source_end": 25,
+        "target_begin": 18,
+        "target_end": 21,
+        "edge": "/kythe/edge/imputes",
+        "source_vname": {
+            "corpus": "corpus",
+            "path": "path/to/Color.java"
+        }
+    }]
+}
+-------------------------------------------------------------------------------
+
+This mapping has 3 `meta` entries. The first entry for `Color`, the second for
+`RED`. Note:
+
+* Each entry doesn't contain names of elements. Each entry contains only
+  position of elements in the source (`Color.java`) and target (`color.js`)
+  files.
+* Each position is defined as byte offset inside file and not as line/column.
+  This is required because in Kythe anchors are defined using byte offsets and
+  not line/column. In this example JavaScript indexer will process this
+  mapping and will need to output *anchor* for `Color.java` and indexer
+  doesn't have access to the `Color.java` file (it has access only to JS
+  files). Because of that JS indexer can't translate line/column to byte
+  offset.
+* Entry doesn't contain vnames of elements in `Color.java` or `color.js` and
+  instead contains positions. Vnames of nodes are internal details of each
+  indexer and subject to change. Generator usually a standalone tool that
+  doesn't know rules for producing vnames for specific language so it's
+  impossible for Generator to output vnames of nodes.
+
+To pass this mapping to the JavaScript Indexer Generator will append it
+as a comment at the last line of `color.js`:
+
+[source,javascript]
+-------------------------------------------------------------------------------
+const Color = {
+  RED: 0,
+};
+
+// Kythe Indexing Metadata:
+// {"type":"kythe0","meta":[{"type":"anchor_anchor","source_begin":13,...
+-------------------------------------------------------------------------------
+
+Inlining metadata inside `color.js` has benefit of avoiding passing extra
+files to Indexer. All Indexer needs is to know that some JavaScript files can
+contain metadata on the last line and parse it.
```

Comment:
For the sake of separation of concerns, the generator would like to stay independent of Kythe, and the indexer of TargetLang would like to stay independent of any SourceLang or metadata.

Indexer-independence can be achieved by the generator emitting separate entries file. But generator independence is also possible, by having stable VName convention in both Source- and TargetLang. Then an external tool could generate the entries itself, using `generates` directly, instead of `imputes`.

Reply:
Stable semantic VNames are infeasible for a substantial fraction of languages and this is intentional.

For some languages with intentionally constrained purposes it is reasonable to document and specify VName construction rules.  This is what was done for protobuf, for example.  It is not feasible to do so for more full-features and evolving languages like C++, Java or JavaScript.

Reply:
Unfortunately even if we had a stable naming convention on both sides (which as @shahms pointed out turns out to be trickier than it seems), we still run into the fact that most code generators don't have enough information to emit complete names for the entities they define. The indexer is able to do this because it has the dependencies resolved already, but a code generator generally works a file at a time and cannot predict how the generated file will be built (e.g., the import path for a Go package, the preprocessor settings and include state for a C translation unit). At best, most generators can give only partial information.

A further complication is that many code generators (most, in my experience) do not work in an abstract model of the output language, but simply write text to a file. The "knowledge" of what program structures are being generated is implicit in the control flow of the program, and while the author of the code generator (hopefully) knows what she is doing, adding instrumentation is therefore fiddly.

So while it's certainly true that using location as a proxy for identity is _also_ not ideal in some ways, it has the advantage that it doesn't require a new abstraction for the language to be built and maintained: Instrumenting the code generator can be done via lightweight location tracking, and all the language-specific knowledge stays in the indexer.

As you might have noted, though, there are some cases where we can do better: For example, protocol buffers already contain a well-defined addressing scheme for structures in the schema language, which are provided to both the code generator and the indexer in the descriptor messages. This works out primarily because the naming scheme is part of the language itself, rather than an addition.

Reply:
Thank you. The incomplete information argument is convincing. So assuming stable language-level VNames, the tradeoff is making the indexer somewhat more complicated (parsing inline data), vs creating an external tool which would need to depend on build-system details. Got it!

Reply:
> [T]he tradeoff is making the indexer somewhat more complicated (parsing inline data), vs. creating an external tool which would need to depend on build-system details.

Exactly. There is no adequate build-system abstraction anyway (since ~all extant build tools except possibly Nix are defined by their implementation), and we'd rather pay the extra cost on the Kythe side than in the code generator.

Reply:
Robin, could you suggest a way to improve this doc to make it clearer? I thought by providing two different approaches/examples and explaining each approach it would make it clear, but probably it didn't. Should I add a section about `imputes` vs `generates`?